dartagnan/src/main/java/com/dat3m/dartagnan/analysis/graphRefinement/graphs/eventGraph/stat/FenceGraph.java

```diff
@@ -0,0 +1,132 @@
+package com.dat3m.dartagnan.analysis.graphRefinement.graphs.eventGraph.stat;
+
+import com.dat3m.dartagnan.analysis.graphRefinement.util.EdgeDirection;
+import com.dat3m.dartagnan.program.Thread;
+import com.dat3m.dartagnan.verification.model.Edge;
+import com.dat3m.dartagnan.verification.model.EventData;
+import com.dat3m.dartagnan.verification.model.ExecutionModel;
+
+import java.util.*;
+import java.util.stream.Stream;
+
+public class FenceGraph extends StaticEventGraph {
+
+    private final String fenceName;
+    private Map<Thread, List<EventData>> threadFencesMap;
+
+    public FenceGraph(String fenceName) {
+        this.fenceName = fenceName;
+    }
+
+    @Override
+    public boolean contains(EventData a, EventData b) {
+        if (a.getThread() != b.getThread() || a.getLocalId() >= b.getLocalId())
+            return false;
+
+        //TODO: We might want to employ binary search instead of linear search
+        return threadFencesMap.get(a.getThread()).stream()
+                .anyMatch(fence -> a.getLocalId() < fence.getLocalId() && fence.getLocalId() < b.getLocalId());
+    }
+
+    @Override
+    public void constructFromModel(ExecutionModel model) {
+        super.constructFromModel(model);
+
+        threadFencesMap = new HashMap<>();
+        for (Thread t : model.getThreads()) {
+            threadFencesMap.put(t, new ArrayList<>());
+        }
+        Set<EventData> fenceEvents = model.getFenceMap().get(fenceName);
+        if (fenceEvents == null) {
+            return;
+        }
+
+        for (EventData fence : fenceEvents) {
+            threadFencesMap.get(fence.getThread()).add(fence);
+        }
+        for (List<EventData> fenceList : threadFencesMap.values()) {
+            fenceList.sort(Comparator.comparingInt(EventData::getLocalId));
```

Comment:
we need to keep them ordered because of `getNextFence()` right?

Reply:
Yes. There is no reason to not sort the events in order of appearance I would say. It makes working with them easier.